app/models/term.rb

```diff
@@ -46,21 +46,58 @@ def set_custom_field(field, value)
     self.custom_fields[field] = value
   end
 
-  def already_exists?
-    errors.added?(:uri, :uniquness) || errors.added?(:uri_hash, :uniqueness)
-  end
-
   private
+    def cast_custom_fields
+      custom_fields.each do |k, v|
+        next if v.nil?
+        raise "custom_field #{k} is not a valid custom field" unless vocabulary.custom_fields.keys.include?(k)
+
+        case vocabulary.custom_fields[k][:data_type]
+        when 'string'
+          raise "custom_field #{k} must be a string" unless v.is_a?(String)
+        when 'integer'
+          if valid_integer?(v)
+            custom_fields[k] = v.to_i if v.is_a?(String)
+          else
+            raise "custom_field #{k} must be an integer"
+          end
+        when 'boolean'
+          if valid_boolean?(v)
+            custom_fields[k] = (v == 'true') ? true : false if v.is_a?(String)
+          else
+            raise "custom_field #{k} must be a boolean"
```

Comment:
```suggestion
            raise "custom_field #{k} must be a boolean. Could not cast String \"#{v}\" to a boolean."
```
How do you feel about this proposed addition?

Reply:
The additions both look good. I'll add them in.